### DISCUSSION TITLE
[Question][Hw1] Internal Structure of an Instruction

### DISCUSSION THREAD

**Discussion by pingpingy1** [prosyslab-classroom/cs424-program-reasoning#387] [Q&A]
I understand from the LLVM Primer slides that most things in LLVM IR are represented as `llvalue` instances.
I also found in the OCaml Llvm module API that the functions `num_operands` and `operand` can be used to access the operands of the instruction if the input `llvalue` instance is in fact an instruction.

However, I seem to be having little luck in finding how the operands are placed and interpreted.
For example, suppose that `Llvm.instr_opcode instr` returned `Llvm.Opcode.Add`.
How many operands should this instruction normally have?
If this was interpreted as the addition arithmetic expression, then I would guess 2.
But if this addition instruction also implied an assignment command implicitly, then I think 3 operands (2 sources, 1 destination) should be provided.

Is there documentation that explains such internal structures of instructions?

Thank you.

**Comment by Hhro** [comment]
Hi,

I think this LLVM language reference can help you:
https://llvm.org/docs/LangRef.html

For instance, if you want to find some information about the 'add' instruction, you can find it at https://llvm.org/docs/LangRef.html#add-instruction.

Thanks.
